Deprecate api-version field in kubeconfig file

janetkuo opened the pull request:
Fixes #14318. 
- Deprecated `api-version` field, and replace it with `preferred-versions` field (defined in pkg/client/unversioned/clientcmd/api/#Cluster.PreferredGroupVersions, type `[]string`)
- Users may specify preferred versions in kubeconfig file (`preferred-versions`) or using kubectl flag (`--preferred-versions`) 
- sample usage: 
  - kubectl: `--preferred-versions=group1/version1 --preferred-versions=group2/version2` or `--preferred-versions=group1/version1,group2/version2`
  - kubeconfig: 
  
  ```
  preferred-versions:
  - group1/version1
  - group2/version2
  ```

cc @lavalamp

janetkuo committed:
Replace api-version field with preferred-versions in kubeconfig file

lavalamp commented on pkg/client/unversioned/clientcmd/api/v1/types.go:
```diff
@@ -54,7 +54,7 @@ type Preferences struct {
 	// Server is the address of the kubernetes cluster (https://hostname:port).
 	Server string `json:"server"`
-	// APIVersion is the preferred api version for communicating with the kubernetes cluster (v1, v2, etc).
+	// DEPRECATED: APIVersion is the preferred api version for communicating with the kubernetes cluster (v1, v2, etc).
```
Awesome-- can you add a note here about why it's deprecated? e.g.

// Because a cluster can run multiple API groups and potentially multiple versions of each, it no longer makes sense to specify a single value for the cluster version. This field isn't really needed anyway, so we're deprecating it without a replacement. It will be ignored if it is present.

JanetKuo replied:
Done

deads2k commented on pkg/client/unversioned/clientcmd/api/v1/types.go:
```diff
@@ -54,7 +54,10 @@ type Preferences struct {
+	// DEPRECATED: APIVersion is the preferred api version for communicating with the kubernetes cluster (v1, v2, etc).
+	//  Because a cluster can run multiple API groups and potentially multiple versions of each, it no longer makes sense to specify
+	//  a single value for the cluster version. This field isn't really needed anyway, so we're deprecating it without a replacement.
+	//  It will be ignored if it is present.
```
This still looks used to me (https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/client_config.go#L101).  Is this a statement of future intent?

lavalamp replied:
Good catch. @JanetKuo can you remove that use in this PR? Thanks!

deads2k replied:
> Good catch. @JanetKuo can you remove that use in this PR? Thanks!

gooracle finds a few refs to chase, that was just the biggest one.

lavalamp replied:
Following up on this, @caesarxuchao convinced me that there's still use for something similar, so the plan now is that @JanetKuo will replace this with a PreferredVersions setting. It will take a list of group/version pairs, and will inform version negotiation. examples:
* "v1,experimental/v1alpha1"
* "experimental/v1alpha1" - this means no preference on the v1 api version, but if you get something from experimental, you must use the v1alpha1 version.

We'll do the same for the command line flag, add a --preferred-versions flag. We'll keep --api-version flag working for requests that use a single resource.

deads2k replied:
> "experimental/v1alpha1" - this means no preference on the v1 api version, but if you get something from experimental, you must use the v1alpha1 version.

I like this idea in general, but I'm not sure of "must" versus "must if available".  I don't think we have to debate it before this changes in, but we currently have "must if available" and I've been pretty pleased with its operation.

I am a little concerned over the data structure choice.  If we placed this is a map, then it would be impossible to have multiple mappings for "experimental".  What happens in this scenario if you have "experimental/v1" and "experimental/v2"?

caesarxuchao replied:
> I like this idea in general, but I'm not sure of "must" versus "must if available". I don't think we have to debate it before this changes in, but we currently have "must if available" and I've been pretty pleased with its operation.

I also think we should "must if avalaible", and let NegotiateVersion report "fall back to XXX version" if the preferred version is not available.

> What happens in this scenario if you have "experimental/v1" and "experimental/v2"?
Will we need to express two preferred versions for one group?

lavalamp replied:
> What happens in this scenario if you have "experimental/v1" and "experimental/v2"?

@JanetKuo and I decided easiest is first setting wins. In the future, we could extend this to be a preference order-- so kubectl would be allowed to use anything in the list, but would use first if possible before trying second & further ones. You can't do that with a map. Also, you can't really express a map on the command line flag, and the behavior should be consistent between them.

> "must" versus "must if available"

Hm, my understanding was that the current flag is "must"? IMO if you're setting the flag, you don't want it to silently not do what you asked-- if what you asked was impossible, don't you want to know about it?

deads2k replied:
> Will we need to express two preferred versions for one group?

I don't think we do and I'd prefer that we prevent that from happening.  Having a simple list of `PreferredVersion` doesn't prevent someone from expressing two preferred versions.  If we used a map, they couldn't do it.

Would you prefer the stronger structure or additional validation during save and use?

deads2k replied:
> Hm, my understanding was that the current flag is "must"? IMO if you're setting the flag, you don't want it to silently not do what you asked-- if what you asked was impossible, don't you want to know about it

I think there's a difference between the flag, the value inside of a resource file, and the value inside of the kubeconfig, but I haven't delved into it too deeply.

lavalamp replied:
I think the flag & config file behave the same, but the value inside an individual resource file is different.

deads2k commented:
Is the associated flag going away too?

deads2k commented:
> Is the associated flag going away too?

I should explain.  I think that the flag still has utility for ambiguous cases, so that a user can explicitly choose which group/version tuple he wants.  I don't feel strongly about a flag in the kubeconfig file, but I don't think I want to lose the power altogether.

lavalamp commented:
@deads2k Yeah, the flag seems moderately more salvageable-- we will either provide a replacement or make it work in a backwards-compatible way.

JanetKuo commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -85,6 +85,7 @@ func NewCmdAnnotate(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 		Long:    annotate_long,
 		Example: annotate_example,
 		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckDeprecatedGloablFlags(cmd)
```
Is there a better way to deprecate global flags other than calling it over and over in each kubectl command?

JanetKuo commented:
It's ready for review again. PTAL. Changes made:
- Added a new field `PreferredVersions` (string) to both clientcmd/api#Cluster and clientcmd/api/v1#Cluster
- Deprecated `APIVersion` of both clientcmd/api#Cluster and clientcmd/api/v1#Cluster
- Added a new global flag `--preferred-versions` to kubectl 
- Deprecated the global flag `--api-version` of kubectl 
- APIVersion of client Config is set to the value of api.Cluster.PreferredVersions, or api.Cluster.APIVersion if PreferredVersions isn't set (i.e., PreferredVersions overrides APIVersion when they're both set).

JanetKuo commented:
@k8s-bot test this please

caesarxuchao commented on pkg/client/unversioned/clientcmd/client_config.go:
```diff
@@ -98,7 +98,12 @@ func (config DirectClientConfig) ClientConfig() (*client.Config, error) {
 		clientConfig.Host = u.String()
 	}
-	clientConfig.Version = configClusterInfo.APIVersion
+	// TODO: clientConfig.Version now contains mutiple versions, we'll need to modify related code for version negotiation
```
In #14592, I'm making client.Config group specific, so clientConfig.Version will only contain one version. Will that cause any conflicts with this PR?

JanetKuo replied:
There's no other conflicts. Per our discussion we'll tackle this after we add a group parameter to this function, i.e., `ClientConfig()`.

JanetKuo commented:
@k8s-bot test this

JanetKuo commented:
@k8s-bot test this

deads2k commented on pkg/client/unversioned/clientcmd/api/v1/types.go:
```diff
@@ -54,8 +54,13 @@ type Preferences struct {
+	//  It will be ignored if it is present.
 	APIVersion string `json:"api-version,omitempty"`
+	// PreferredVersions takes a string of multiple group/version pairs (separated by comma), and informs version negotiation
+	PreferredVersions string `json:"preferred-versions,omitempty"`
```
Can we call this `PreferredGroupVersions` to avoid ambiguity and make the struct field `[]string`?  That makes the code more obvious.  I don't object to the shorter flag name, but there could be usability benefits to setting it to be a `[]string` flag as well.

JanetKuo replied:
Thanks @deads2k. I can see the value of using `[]string` instead of `string`. But users would prefer to type `--preferred-versions=g1/v1,g2/v2` rather than  `--preferred-versions=g1/v1 --preferred-versions=g2/v2`, imo. 

@lavalamp @caesarxuchao what do you think?

liggitt replied:
it should be structured internally. the StringList CLI binding allows both the `--foo=bar -foo=baz` and `--foo=bar,baz` forms

JanetKuo replied:
Thanks I'll change this as advised.

deads2k commented on pkg/client/unversioned/clientcmd/overrides.go:
```diff
@@ -172,6 +175,7 @@ func BindAuthInfoFlags(authInfo *clientcmdapi.AuthInfo, flags *pflag.FlagSet, fl
 func BindClusterFlags(clusterInfo *clientcmdapi.Cluster, flags *pflag.FlagSet, flagNames ClusterOverrideFlags) {
 	flagNames.APIServer.BindStringFlag(flags, &clusterInfo.Server)
 	flagNames.APIVersion.BindStringFlag(flags, &clusterInfo.APIVersion)
+	flagNames.PreferredVersions.BindStringFlag(flags, &clusterInfo.PreferredVersions)
```
You can have `[]string` flags now that would allow `--preferred-version=foo --preferred-version=bar`.  What about `--version`?  Too generic?

deads2k commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -225,6 +225,17 @@ func GetFlagString(cmd *cobra.Command, flag string) string {
 	return s
 }
 
+func CheckDeprecatedGlobalFlags(cmd *cobra.Command) {
```
I thought that when I used a deprecated flag, `cobra` automatically notified me.  Is that not the case?

JanetKuo replied:
Thanks. It's now removed and I use `MarkDeprecated` in overrides.go (when binding flags) instead.

caesarxuchao commented:
Just a heads-up, without fixing how we load Cluster information into client.Config and other issues with client.Config (#14592), this PR may cause wrong behavior if user specifies multiple versions in the PreferredVersions in the kubeconfig. We probably should merge this after v1.1.

JanetKuo commented:
@k8s-bot test this

JanetKuo commented:
@deads2k @liggitt I renamed `PreferredVersions` to `PreferredGroupVersions`, changed its type and `--preferred-versions` flag's type from `string` to `[]string`. 

Note that in kubeconfig and kubectl we use `preferred-versions`, instead of something like `preferred-group-versions`.

JanetKuo commented:
Not sure why @k8s-merge-robot mark this as size/XS? cc @lavalamp

deads2k commented:
> Not sure why @k8s-merge-robot mark this as size/XS? cc @lavalamp

Oh please tell me the bot now excludes generated files.  That would be great.

deads2k commented on pkg/client/unversioned/clientcmd/overrides.go:
```diff
@@ -171,7 +188,10 @@ func BindAuthInfoFlags(authInfo *clientcmdapi.AuthInfo, flags *pflag.FlagSet, fl
+	// TODO: remove --api-version flag on or after Oct. 5, 2016
 	flagNames.APIVersion.BindStringFlag(flags, &clusterInfo.APIVersion)
+	flagNames.PreferredGroupVersions.BindStringSliceFlag(flags, &clusterInfo.PreferredGroupVersions)
+	flags.MarkDeprecated(FlagAPIVersion, fmt.Sprintf("use --%s instead", FlagPreferredGroupVersions))
```
This needs to be conditional based on whether the `longName` for the `APIVersion` flag is non-empty.  Otherwise that flag may not be bound.

JanetKuo replied:
Fixed

deads2k commented on pkg/kubectl/cmd/config/create_cluster.go:
```diff
@@ -31,13 +31,15 @@ import (
+	name         string
+	server       util.StringFlag
+	apiVersion   util.StringFlag
+	//	preferredVersions     util.StringFlag
```
cruft

JanetKuo replied:
Removed

deads2k commented on pkg/kubectl/cmd/config/create_cluster.go:
```diff
@@ -118,6 +124,11 @@ func (o *createClusterOptions) modifyCluster(existingCluster clientcmdapi.Cluste
 		modifiedCluster.Server = o.server.Value()
 	}
+	if o.preferredGroupVersions.Provided() {
+		fmt.Printf("preferred versions provided: o.preferredGroupVersions = %#v, o.preferredGroupVersions.Value() = %v\n", o.preferredGroupVersions, o.preferredGroupVersions.Value())
```
debugging lines?

JanetKuo replied:
Yes. Removed now.

deads2k commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -90,6 +90,7 @@ func NewCmdEdit(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	usage := "Filename, directory, or URL to file to use to edit the resource"
 	kubectl.AddJsonFilenameFlag(cmd, &filenames, usage)
 	cmd.Flags().StringP("output", "o", "yaml", "Output format. One of: yaml|json.")
+	// TODO: change the description to "[...] (default preferred-versions)"?
```
Yes, I think so.

JanetKuo replied:
Updated

deads2k commented:
minor comments.  lgtm otherwise.

bgrant0607 commented:
When making a change of this size, the PR really needs an explanation.

JanetKuo commented:
@deads2k comment addressed, thanks! @bgrant0607 PR description updated.

bgrant0607 commented:
cc @jlowdermilk

jlowdermilk commented on pkg/client/unversioned/clientcmd/api/types.go:
```diff
@@ -56,8 +56,13 @@ type Cluster struct {
+	//  a single value for the cluster version. Deprecate replace it with PreferredGroupVersions. It will be ignored if it is present.
 	APIVersion string `json:"api-version,omitempty"`
+	// PreferredGroupVersions takes a string slice of multiple group/version pairs (separated by comma), and informs version negotiation.
+	// It contains preferred group/versions for communicating with the kubernetes cluster (g1/v1, g2/v2, etc).
```
The comment says the entries should be group/version pairs, but none of your tests do this, and it doesn't look like the loading code splits either. Unless there is a specific reason to make entries be group/version, can they just be plain version strings?

lavalamp replied:
Is the comment mistaken? No, it is the tests that are wrong. This *must* be group/version pairs. Versions no longer mean anything without an associated group. "v1" is special and refers to the legacy v1 api.

JanetKuo replied:
The test is for `""` group so the group part is missing; I'll add tests for group/version and make it more clear. `PreferredGroupVersions` be used later for creating client.Config's group/version (not yet implemented). Having plain version strings isn't enough since version doesn't make sense without its group and `Cluster` doesn't have group information.

JanetKuo replied:
Fixed tests.

lavalamp replied:
Thanks, much better!

lavalamp commented on pkg/client/unversioned/clientcmd/api/types_test.go:
```diff
@@ -43,15 +43,15 @@ func ExampleOfOptionsConfig() {
-		APIVersion:            "v1beta1",
-		InsecureSkipTLSVerify: false,
+		Server:                 "https://bravo.org:8080",
+		PreferredGroupVersions: []string{"v1beta1"},
```
Please put in entries that exercise both cases-- "v1", "experimental/v1alpha1"

JanetKuo replied:
Done

lavalamp commented on pkg/client/unversioned/clientcmd/client_config_test.go:
```diff
@@ -33,8 +33,8 @@ func createValidTestConfig() *clientcmdapi.Config {
-		Server:     server,
-		APIVersion: testapi.Default.Version(),
+		Server:                 server,
+		PreferredGroupVersions: []string{testapi.Default.Version()},
```
GroupVersion()!

JanetKuo replied:
Done

deads2k commented:
lgtm on green.

caesarxuchao commented:
@deads2k, I don't think we can merge this now. I haven't followed up on this PR, but last time I looked at it, there is no logic to parse the multiple groupVersions in kubeconfig yet, so if people specify multiple groupVersions, it may cause unexpected behavior. 

I will continue my work on client.Config (#14592) which defines how to parse the preferred versions, then this PR can be merged.

JanetKuo commented:
I agree. This PR will wait for #14592 to be merged.

JanetKuo commented:
Based on https://github.com/kubernetes/kubernetes/pull/15874#issuecomment-149353809, should we change it to `--preferred-versions=version1.group1,version2.group2`, or we don't need --preferred-versions at all?

bgrant0607 commented:
Is this waiting on anything? If it's going to be in 1.1, it needs to be finished ASAP.

janetkuo commented:
@bgrant0607 this is waiting for the refactor of unversioned.Config, and won't going to be in 1.1. Removing the cherry-pick tag.

bgrant0607 commented:
@lavalamp Is this still waiting on anything?

lavalamp commented on pkg/util/string_slice_flag.go:
```diff
@@ -0,0 +1,64 @@
+limitations under the License.
+*/
+
+package util
```
Can we put this in `pkg/util/flagtypes` or something like that? Just to avoid adding more stuff to util.

lavalamp commented:
No, I think this should be OK for 1.2. Sorry, @janetkuo for totally forgetting about this. Can you see how hard it is to rebase this now?

nikhiljindal commented:
I remember this PR being mentioned in one of our meetings as required in 1.2, but I dont see the 1.2 milestone on it.

I can take this up if @janetkuo is busy and this is required for 1.2

nikhiljindal commented:
Discussed this with @lavalamp in our meeting today.
We definitely want to deprecate api-versions flag in 1.2.

Adding the preferred-versions flag is not critical.
Once https://github.com/kubernetes/kubernetes/pull/21272 is merged, users can mention resource.group in kubectl commands to disambiguate resources in 2 groups.

@janetkuo Can you update this PR to just deprecate api-versions flag? I can take it up incase you are busy with other 1.2 tasks.

janetkuo commented:
> @janetkuo Can you update this PR to just deprecate api-versions flag? 

@nikhiljindal Yes I'll update this.

bgrant0607 commented:
Yes, @nikhiljindal please take this over.
